### ISSUE TITLE
Basic support for efficient cross-validation of GP models

### ISSUE THREAD

**Issue by Balandat** [cornellius-gp/gpytorch#130]
It would be nice to have some basic support for efficient cross-validation of GP models.

A standard way to do this for k-Fold CV with small fold size is to keep the parameters of the model fixed, update the training input and targets with the k-th fold removed, and then predict the k-th fold with that model. This works, but in a naive implementation we're potentially leaving a lot of useful information on the table by not using the information from the prediction of the full model. 

A likely much more efficient way is, for the k-th fold, to pass the elements of the solution of the full linear_cg problem corresponding to the complement of the k-th fold as initial_guess into linear_cg.

Does this sound reasonable? (I'm not sure how to adopt this to the sparse GP methods though...)

**Comment by Balandat** [comment]
cc @darbour

**Comment by jacobrgardner** [comment]
Something to consider: GPyTorch is intended to operate fully in batch mode. This means that if I have training data that looks like `k x n x d`, training labels that look like `k x n`, and test points that look like `k x t x d`, GPyTorch should handle making predictions for k independent GPs in a batched parallel fashion, where the $ith$ GP uses the ith batch of training data and tests on the ith batch of test data.

The reason I mention this is that k-fold CV is often used when not enough data is available to just split off a validation set. In these settings, the parallel compute capability of a decent GPU might be enough that you can just do all the folds at once in parallel in time not much larger than doing a single fold.

**Comment by Balandat** [comment]
@jacobrgardner, I'm having trouble getting the batched training to work, b/c the mean output of the model seems off. Here is the simple_gp_regression example for two identical batches (`k=2`, `n=11`, `d=1`):

```
from torch.autograd import Variable
train_x = Variable(torch.linspace(0, 1, 11)).view(1, -1, 1).repeat(2, 1, 1)
train_y = (
    Variable(torch.sin(train_x.data * (2 * math.pi))
             + torch.randn(train_x.size()) * 0.2)
).view(1, -1, 1).repeat(2, 1, 1)

likelihood = GaussianLikelihood(log_noise_bounds=(-5, 5))
model = ExactGPModel(train_x.data, train_y.data, likelihood)

# Find optimal model hyperparameters
model.train()
likelihood.train()

# Use the adam optimizer
optimizer = torch.optim.Adam([
    {'params': model.parameters()},  # Includes GaussianLikelihood parameters
], lr=0.1)

# "Loss" for GPs - the marginal log likelihood
mll = gpytorch.mlls.ExactMarginalLogLikelihood(likelihood, model)

training_iter = 50
for i in range(training_iter):
    # Zero gradients from previous iteration
    optimizer.zero_grad()
    # Output from model
    output = model_batch(train_x)
    # Calc loss and backprop gradients
    loss = -mll(output, train_y)
    loss.backward()
    print('Iter %d/%d - Loss: %.3f   log_lengthscale: %.3f   log_noise: %.3f' % (
        i + 1, training_iter, loss.data[0],
        model.covar_module.log_lengthscale.data[0, 0],
        model.likelihood.log_noise.data[0]
    ))
    optimizer.step()

---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-83-06d95aee3890> in <module>()
     28     output = model_batch(train_x)
     29     # Calc loss and backprop gradients
---> 30     loss = -mll(output, train_y)
     31     loss.backward()
     32     print('Iter %d/%d - Loss: %.3f   log_lengthscale: %.3f   log_noise: %.3f' % (

/data/users/balandat/fbsource/fbcode/buck-out/dev-nosan/gen/experimental/ae/bento_kernel_ae_experimental#link-tree/gpytorch/module.py in __call__(self, *inputs, **kwargs)
    182 
    183     def __call__(self, *inputs, **kwargs):
--> 184         outputs = self.forward(*inputs, **kwargs)
    185         if (
    186             isinstance(outputs, Variable)

/data/users/balandat/fbsource/fbcode/buck-out/dev-nosan/gen/experimental/ae/bento_kernel_ae_experimental#link-tree/gpytorch/mlls/exact_marginal_log_likelihood.py in forward(self, output, target)
     38         # Get log determininat and first part of quadratic form
     39         inv_quad, log_det = covar.inv_quad_log_det(
---> 40             inv_quad_rhs=(target - mean).unsqueeze(-1), log_det=True
     41         )
     42 

/data/users/balandat/fbsource/fbcode/buck-out/dev-nosan/gen/experimental/ae/bento_kernel_ae_experimental#link-tree/gpytorch/lazy/lazy_variable.py in inv_quad_log_det(self, inv_quad_rhs, log_det)
    404                     )
    405             elif self.ndimension() > 3 or inv_quad_rhs.ndimension() > 3:
--> 406                 raise RuntimeError
    407 
    408         matrix_size = self.size(-1)

RuntimeError: 
```

The issue appears to be that the model evaluated on the training data yields a tensor with `k` elements (rather than `k x n`). Am I doing something wrong here?

**Comment by jacobrgardner** [comment]
I think Sam Stanton here ran in to a similar issue in the past couple days and has a branch on a fork that fixes this.

Let me ping him.

**Comment by jacobrgardner** [comment]
@Balandat this should hopefully be pretty straight forward now with #158!
